File: Cpp+Python/DailyLeetcode/May/May4.cc
#include <vector>
using namespace std;

class Solution {
public:
  int numRescueBoats(vector<int> &people, int limit) {
    int numBoats = 0;
    sort(people.begin(), people.end());
    // want to pair heaviest with lightest when possible

    int l = 0, r = people.size() - 1;
    while (l < r) {
      if (people[l] + people[r] <= limit) {
        numBoats++;
        l++;
        r--;
      } else {
        numBoats++;
        r--;
      }
      if (l == r) {
        numBoats++;
      }
    }
    return numBoats;
  }
};

File: go_sols/DailyLeetcode/May/May4.go
package May

import (
	"sort"
)

func numRescueBoats(people []int, limit int) int {
	result := 0
	sort.Ints(people)
	l, r := 0, len(people)-1

	// 1 2 2 3
	for l < r {
		if people[l]+people[r] <= limit {
			l++
		}
		r--
		result++

		if l == r {
			result++
		}
	}
	return result
}
